runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/DedupingOperator.java

```diff
@@ -0,0 +1,184 @@
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.beam.runners.flink.translation.wrappers.streaming.state.KeyGroupCheckpointedOperator;
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.Coder.Context;
+import org.apache.beam.sdk.coders.VarIntCoder;
+import org.apache.beam.sdk.util.ValueWithRecordId;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.runtime.state.KeyGroupStatePartitionStreamProvider;
+import org.apache.flink.runtime.state.KeyGroupsList;
+import org.apache.flink.runtime.state.KeyedStateBackend;
+import org.apache.flink.runtime.state.KeyedStateCheckpointOutputStream;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.joda.time.Duration;
+
+/**
+ * Remove values with duplicate ids.
+ */
+public class DedupingOperator<T> extends AbstractStreamOperator<WindowedValue<T>>
+    implements OneInputStreamOperator<WindowedValue<ValueWithRecordId<T>>, WindowedValue<T>>,
+    KeyGroupCheckpointedOperator {
+
+  private static final long MAX_RETENTION_SINCE_ACCESS = Duration.standardMinutes(10L).getMillis();
+  private static final long MAX_CACHE_SIZE = 100_000L;
+
+  private transient Map<Integer, LoadingCache<ByteBuffer, AtomicBoolean>> dedupingCache;
+  private transient KeyedStateBackend<ByteBuffer> keyedStateBackend;
+
+  @Override
+  public void open() throws Exception {
+    super.open();
+    dedupingCache = new HashMap<>();
+    keyedStateBackend = getKeyedStateBackend();
+  }
+
+  private static class TrueBooleanLoader extends CacheLoader<ByteBuffer, AtomicBoolean> {
+    @Override
+    public AtomicBoolean load(ByteBuffer key) throws Exception {
+      return new AtomicBoolean(true);
+    }
+  }
+
+  @Override
+  public void processElement(
+      StreamRecord<WindowedValue<ValueWithRecordId<T>>> streamRecord) throws Exception {
+    ByteBuffer currentKey = keyedStateBackend.getCurrentKey();
+    int groupIndex = keyedStateBackend.getCurrentKeyGroupIndex();
+    if (shouldOutput(groupIndex, currentKey)) {
+      WindowedValue<ValueWithRecordId<T>> value = streamRecord.getValue();
+      output.collect(streamRecord.replace(value.withValue(value.getValue().getValue())));
```

Comment:
`value.withValue(value.getValue().getValue())` might be one of my favorite pieces of code ever :-) especially since it is just right. If `StreamRecord` had a similar API we could go ever further...